File: mobile/src/utils/appVersion.ts
/**
 * Tell the server which build of the app is calling, and hear when it is too old.
 *
 * Mobile builds can't be rolled back and the API has no version negotiation,
 * so an old build with a known defect — like the check-in queue fixed in
 * September 2026 — keeps running until the officer happens to update. The
 * server compares `X-App-Build` with a configured minimum and answers 426 when
 * a build is below it; this module sends the headers and raises the
 * "update required" signal the app shell turns into a blocking message.
 */
import axios from 'axios';
import Constants from 'expo-constants';
import { Platform } from 'react-native';

function buildNumber(): string {
  const config = Constants.expoConfig;
  const value = Platform.OS === 'ios' ? config?.ios?.buildNumber : config?.android?.versionCode;
  return value != null ? String(value) : '0';
}

export const APP_BUILD = buildNumber();

export function appVersionHeaders(): Record<string, string> {
  return { 'X-App-Platform': Platform.OS, 'X-App-Build': APP_BUILD };
}

export const UPDATE_REQUIRED_STATUS = 426;

type Listener = () => void;
const listeners = new Set<Listener>();

/** Subscribe to "this build is too old". Returns an unsubscribe function. */
export function onUpdateRequired(listener: Listener): () => void {
  listeners.add(listener);
  return () => listeners.delete(listener);
}

export function notifyUpdateRequired(): void {
  for (const listener of listeners) {
    try {
      listener();
    } catch {
      // A listener's failure must not break the request path.
    }
  }
}

let installed = false;

/** Add the headers and the 426 check to every axios request. Call once at start-up. */
export function installAppVersionOnAxios(): void {
  if (installed) return;
  installed = true;
  Object.assign(axios.defaults.headers.common, appVersionHeaders());
  axios.interceptors.response.use(undefined, (error) => {
    if (error?.response?.status === UPDATE_REQUIRED_STATUS) notifyUpdateRequired();
    return Promise.reject(error);
  });
}

File: backend/api/middleware/app_version.py
"""
Refuse mobile builds below a configured minimum (AUDIT-2026-09-17, Phase 2F).

Mobile builds can't be rolled back, so a build with a known defect keeps
running until the officer happens to update. The app sends `X-App-Platform`
and `X-App-Build` on every request (`mobile/src/utils/appVersion.ts`); below
`MIN_APP_BUILD[platform]` this answers 426 `app_update_required`, which the
app shows as a blocking "Update required" message.

A request without the headers passes: the web admin sends none, and neither do
builds from before this gate. Those older builds are the ones whose offline
queue loses a check-in that fails, so refusing them would make things worse;
they are retired by updating devices, not by the server.

With no minimum configured (the default) nothing is refused.
"""
import logging

from django.conf import settings
from django.http import JsonResponse

logger = logging.getLogger(__name__)

#: Paths never refused: monitoring must work whatever build asks.
EXEMPT_PREFIXES = ('/api/v1/health/',)


class MinimumAppBuildMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        refusal = self._refusal(request)
        return refusal if refusal is not None else self.get_response(request)

    def _refusal(self, request):
        platform = request.headers.get('X-App-Platform', '').strip().lower()
        raw_build = request.headers.get('X-App-Build', '').strip()
        if not platform or not raw_build or request.path.startswith(EXEMPT_PREFIXES):
            return None
        minimum = (getattr(settings, 'MIN_APP_BUILD', None) or {}).get(platform) or 0
        try:
            build = int(raw_build)
        except ValueError:
            return None  # a dev build or a typo: not ours to judge
        if build >= minimum:
            return None
        logger.info("Refused %s build %s (minimum %s): %s", platform, build, minimum, request.path)
        return JsonResponse(
            {
                'error': 'app_update_required',
                'detail': 'This version of the app is no longer supported. '
                          'Install the latest version to continue.',
                'minimum_build': minimum,
            },
            status=426,
        )
